src/core/internal/arrayop.d

```diff
@@ -0,0 +1,422 @@
+        else static if (is(T == double))
+            return __simd(XMM.LODUPD, *cast(const vec*) p);
+        else
+            return __simd(XMM.LODDQU, *cast(const vec*) p);
+    }
+
+    __vector(T[N]) binop(string op, T, size_t N)(in __vector(T[N]) a, in __vector(T[N]) b)
+    {
+        pragma(inline, true);
+        return mixin("a " ~ op ~ " b");
+    }
+
+    __vector(T[N]) unaop(string op, T, size_t N)(in __vector(T[N]) a)
+            if (op[0] == 'u')
+    {
+        pragma(inline, true);
+        return mixin(op[1 .. $] ~ "a");
+    }
+}
+
+// mixin gen
+
+// filter out ops without matching SSE/SIMD instructions (could be composed of several instructions though)
+template opsSupported(bool fail, T, ops...) if (ops.length > 1)
+{
+    enum opsSupported = opsSupported!(fail, T, ops[0 .. $ / 2])
+            && opsSupported!(fail, T, ops[$ / 2 .. $]);
+}
+
+template opsSupported(bool fail, T, string op)
+{
+    static if (isUnaryOp(op))
+    {
+        enum opsSupported = is(typeof((T a) => mixin(op[1 .. $] ~ " a")));
+        static assert(!fail || opsSupported,
+                "Unary op `" ~ op[1 .. $] ~ "` not supported for element type " ~ T.stringof ~ ".");
+    }
+    else
+    {
+        enum opsSupported = is(typeof((T a, T b) => mixin("a " ~ op ~ " b")));
+        static assert(!fail || opsSupported,
+                "Binary op `" ~ op ~ "` not supported for element type " ~ T.stringof ~ ".");
+    }
+}
+
+// filter out things like float[] = float[] / size_t[]
+enum compatibleVecTypes(E, T : T[]) = is(Unqual!T == Unqual!E); // array elem types must be same (maybe add cvtpi2ps)
+enum compatibleVecTypes(E, T) = is(T : E); // scalar must be convertible to target elem type
+enum compatibleVecTypes(E, Types...) = compatibleVecTypes!(E, Types[0 .. $ / 2])
+        && compatibleVecTypes!(E, Types[$ / 2 .. $]);
+
+version (GNU_OR_LDC)
+{
+    // leave it to the auto-vectorizer
+    enum vectorizeable(E : E[], Args...) = false;
+}
+else
+{
+    template vectorizeable(E : E[], Args...)
+    {
+        static if (is(vec!E))
+            enum vectorizeable = opsSupported!(false, vec!E, Filter!(not!isType, Args))
+                    && compatibleVecTypes!(E, Filter!(isType, Args));
+        else
+            enum vectorizeable = false;
+    }
+
+    version (X86_64) unittest
+    {
+        static assert(vectorizeable!(double[], const(double)[], double[], "+", "="));
+        static assert(!vectorizeable!(double[], const(ulong)[], double[], "+", "="));
+    }
+}
+
+bool isUnaryOp(string op)
+{
+    return op[0] == 'u';
+}
+
+bool isBinaryOp(string op)
+{
+    if (op.length != 1)
+        return false;
+    switch (op[0])
+    {
+    case '+', '-', '*', '/', '%', '|', '&', '^':
+        return true;
+    default:
+        return false;
+    }
+}
+
+bool isBinaryAssignOp(string op)
+{
+    return op.length == 2 && op[1] == '=' && isBinaryOp(op[0 .. 1]);
+}
+
+string scalarExp(Args...)()
```

Comment:
Desperately needs documentation - for example, what is the format of the RPN string? What are the Args? Where does the RPN string come from?